src/EventListener/KernelRequestListener.php

```diff
@@ -0,0 +1,394 @@
+        private readonly int                 $lifetime,
+        private readonly string              $storageKey,
+        private readonly bool                $considerDnt
+    ){}
+
+    /**
+     * @param RequestEvent $event
+     * @return void
+     */
+    public function onKernelRequest(RequestEvent $event): void
+    {
+        $request = $event->getRequest();
+
+        // Also handles CookieBar not being activated within rootPage
+        if (
+            $this->objRootPage instanceof PageModel ||
+            !$this->scopeMatcher->isFrontendRequest($request) ||
+            !$request->attributes->has('pageModel')
+        )
+        {
+            return;
+        }
+
+        $pageModel = $request->attributes->get('pageModel');
+
+        if ($pageModel instanceof PageModel)
+        {
+            $rootPageObject = PageModel::findByPk($pageModel->rootId);
+
+            if ($rootPageObject instanceof PageModel)
+            {
+                $this->objRootPage = $rootPageObject;
+                $this->objPage = $pageModel;
+
+                $this->prepareCookieBar();
+            }
+        }
+    }
+
+    /**
+     * @param ResponseEvent $event
+     * @return void
+     */
+    public function onKernelResponse(ResponseEvent $event): void
+    {
+        $request = $event->getRequest();
+
+        // Ajax-Request also could render HTML-Output. So it could be that it is detected as PageTemplate.
+        // So normally Ajax-Request does not have pageModel as attribute. Therefore, check the parameter
+        if (
+            !$this->cookiebarModel instanceof CookiebarModel ||
+            !$this->objPage instanceof PageModel ||
+            !$this->scopeMatcher->isFrontendRequest($request) ||
+            !$request->attributes->has('pageModel')
+        )
+        {
+            return;
+        }
+
+        $response = $event->getResponse();
+        $content = $response->getContent();
+
+        if ($this->isPageTemplate($event) === true)
+        {
+            $content = match ($this->cookiebarModel->position)
+            {
+                'bodyAboveContent' => preg_replace("/<body([^>]*)>(.*?)<\/body>/is", "<body$1>$this->rootPageBuffer$2</body>", $content),
+                default => str_replace("</body>", "$this->rootPageBuffer</body>", $content),
+            };
+
+            $content = $this->injectGlobalJs($content);
+            $response->setContent($content);
+
+        }
+
+        if ($request->attributes->has('contentModel'))
+        {
+            $contentModel = $request->attributes->get('contentModel');
+
+            if (!$contentModel instanceof ContentModel)
+            {
+                $contentModel = ContentModel::findByPk($contentModel);
+            }
+
+            // renew $content because using insertTags in modules it could be that contentModel and moduleModel is set
+            $content = $this->parseTemplates($contentModel->typePrefix . $contentModel->type, $content);
```

Comment:
I am not sure this will work if templates variants are in use, i.e. either:

* The fragment uses a template variant, or
* an "iframe type" has been registered only on a template variant.

Reply:
For example, this would not work anymore:

```yaml
contao_cookiebar:
    iframe_types:
        vendortype:
            - content_element/html/my_google_map
            - ce_html_my_google_map
```

as the current code would only consider `content_element/html` or `ce_html` respectiely. In general a `kernel.response` listener is not suitable for this purpose as you lose the information of which template was actually rendered.

Furthermore a subrequest can render multiple different templates.